File: 2176-parallel-courses-iii/parallel-courses-iii.py
from collections import deque, defaultdict
class Solution(object):
    def toposort(self,n,max_time,indegree,time,adj):
        que = deque()
        for i in range(1,n+1):
            if indegree[i]==0:
                que.append(i)
                max_time[i] = time[i-1]

        while que:
            curr_course = que.popleft()
            for next_course in adj[curr_course]:
                max_time[next_course] = max(max_time[next_course],max_time[curr_course]+time[next_course-1])
                indegree[next_course]-=1
                if indegree[next_course]==0:
                    que.append(next_course)         
    def minimumTime(self, n, relations, time):
        adj = defaultdict(list)
        indegree = [0]*(n+1)

        for prev_course,next_course in relations:
            adj[prev_course].append(next_course)
            indegree[next_course]+=1

        maxTime = [0]*(n+1)
        self.toposort(n,maxTime,indegree,time,adj)
        return max(maxTime)
        

File: 2176-parallel-courses-iii/parallel-courses-iii.cpp
class Solution {
public:
    void topoSort(int n,vector<int>&maxTime,vector<int>&indegree,vector<int>& time,unordered_map<int,vector<int>>&adj){
        queue<int>q;
        for(int i=1;i<=n;i++){
            if(indegree[i]==0){
                q.push(i);
                maxTime[i]= time[i-1];
            }
        }

        while(!q.empty()){
            int currCourse = q.front();
            q.pop();

            for (int &next:adj[currCourse]){
                maxTime[next]=max(maxTime[next],maxTime[currCourse]+time[next-1]);
                indegree[next]--;
                if (indegree[next]==0){
                    q.push(next);
                }
            }
        }
    }
    int minimumTime(int n, vector<vector<int>>& relations, vector<int>& time) {
        unordered_map<int,vector<int>>adj;
        vector<int>indegree(n+1,0);
        
        for(vector<int>&rel:relations){
            int prev= rel[0];
            int next = rel[1];
            adj[prev].push_back(next);
            indegree[next]++;
        }
        vector<int>maxTime(n+1,0);
        topoSort(n,maxTime,indegree,time,adj);
        int result = 0;
        for(int i=1;i<=n;i++){
            result = max(result,maxTime[i]);
        }
        return result;
    }
};